### PR TITLE
feat(optimizer)!: Annotate snowflake MD5 functions

### PR DESCRIPTION
**Pull request by fivetran-amrutabhimsenayachit**
Annotate all MD5 functions.

Documentation: https://docs.snowflake.com/en/sql-reference/functions/md5

### CODE REVIEW

**Comment by fivetran-amrutabhimsenayachit** [comment]
`MD5` already has a class in `expression.py`. For other MD5 functions, we can also add classes in `expression.py` and extend them in the `TYPE_TO_EXPRESSIONS` of `snowflake` dialect. But not sure how it may affect the other dialects, hence limited the annotation to `snowflake` dialect.

File: sqlglot/dialects/snowflake.py [hunk 1]
```diff
@@ -493,6 +493,28 @@ def _annotate_reverse(self: TypeAnnotator, expression: exp.Reverse) -> exp.Rever
     return expression
 
 
+def _annotate_anonymous_md5_functions(
+    self: TypeAnnotator, expression: exp.Anonymous
+) -> exp.Anonymous:
+    self._annotate_args(expression)
```
Comment (georgesittas) [inline sqlglot/dialects/snowflake.py]: This isn't quite right. If a function is not being parsed into a non-`Anonymous` expression, we should introduce a new `Func` subclass to represent it and then annotate that new function.

File: sqlglot/dialects/snowflake.py [hunk 2]
```diff
@@ -515,13 +537,20 @@ class Snowflake(Dialect):
...
             exp.Replace,
             exp.Space,
             exp.Uuid,
         },
+        exp.DataType.Type.BINARY: {
+            exp.MD5Digest,
+        },
+        exp.DataType.Type.BIGINT: {
+            *Dialect.TYPE_TO_EXPRESSIONS[exp.DataType.Type.BIGINT],
+        },
```
Comment (georgesittas) [inline sqlglot/dialects/snowflake.py]: Is this necessary?

File: sqlglot/dialects/snowflake.py (hunk 1 above)
Comment (geooo109) [reply sqlglot/dialects/snowflake.py]: To add something here, when we add a new `Func`, we should check other dialects for similar names in order to determine the best representation of this function. For example, [BigQuery REGEXP_INSTR](https://cloud.google.com/bigquery/docs/reference/standard-sql/string_functions#regexp_instr) vs [Snowflake REGEXP_INSTR](https://docs.snowflake.com/en/sql-reference/functions/regexp_instr), here Snowflake supports the arguments of BQ's + some extra ones.

File: sqlglot/dialects/snowflake.py (hunk 2 above)
Comment (fivetran-amrutabhimsenayachit) [reply sqlglot/dialects/snowflake.py]: I annotated the new MD5 functions here.

File: sqlglot/expressions.py [hunk 3]
```diff
@@ -6951,6 +6951,16 @@ class MD5Digest(Func):
...
 
+# Represents the variant of the MD5 function that returns the lower 64 bits as a number
+class MD5NumberLower64(Func):
+    _sql_names = ["MD5_NUMBER_LOWER64"]
+
+
+# Represents the variant of the MD5 function that returns the upper 64 bits as a number
+class MD5NumberUpper64(Func):
+    _sql_names = ["MD5_NUMBER_UPPER64"]
+
```
Comment (geooo109) [inline sqlglot/expressions.py]: Are these functions only used in the Snowflake dialect? 
1. We tend to use a link from the dialect's docs as a comment, instead of explaining with text.
2. We should add some tests in the `test_snowflake.py` using `validate_identity` in order to ensure the correct parsing/generation.
3. Here you should just `pass` (remove `_sql_names`)

**Review by geooo109 — COMMENTED** [review]
Revert the approval needs more changes, see my comment.

File: sqlglot/expressions.py (hunk 3 above)
Comment (fivetran-amrutabhimsenayachit) [reply sqlglot/expressions.py]: Yes, these are used only in Snowflake dialect.
Noted.

**Review by geooo109 — APPROVED** [review]
LGTM!

File: sqlglot/dialects/snowflake.py [hunk 4]
```diff
@@ -1352,6 +1363,9 @@ class Generator(generator.Generator):
             ),
             exp.SHA: rename_func("SHA1"),
             exp.MD5Digest: rename_func("MD5_BINARY"),
+            exp.MD5: rename_func("MD5_HEX"),
```
Comment (geooo109) [inline sqlglot/dialects/snowflake.py]: We can always generate `MD5` when the input is `MD5_HEX` 
The current approach is correct and mergable, do you think we should change this @georgesittas ?

Comment (georgesittas) [reply sqlglot/dialects/snowflake.py]: Yes, I'm in favor of generating the simpler `MD5` variant instead of the `MD5_HEX` alias.

File: tests/dialects/test_snowflake.py [hunk 5]
```diff
@@ -3116,3 +3116,17 @@ def test_bit_aggs(self):
             for name in bit_func:
                 with self.subTest(f"Testing Snowflakes {name}"):
                     self.validate_identity(f"{name}(x)", f"{bit_func[0]}(x)")
+
+    def test_md5_functions(self):
+        self.validate_identity("MD5(col)", "MD5_HEX(col)")
+        self.validate_identity("MD5_HEX(col)")
```
Comment (georgesittas) [inline tests/dialects/test_snowflake.py]: So these tests should be reversed: `MD5` should remain as is, whereas `MD5_HEX` maps to `MD5`.

File: tests/dialects/test_snowflake.py [hunk 6]
```diff
@@ -3116,3 +3116,17 @@ def test_bit_aggs(self):
...
+        self.validate_identity("MD5_BINARY(col)")
+        self.validate_identity("MD5_NUMBER_LOWER64(col)")
+        self.validate_identity("MD5_NUMBER_UPPER64(col)")
+
+        self.validate_identity(
+            "SELECT MD5_HEX('test'), MD5_BINARY('test'), MD5_NUMBER_LOWER64('test'), MD5_NUMBER_UPPER64('test')"
+        )
+        self.validate_identity(
+            "SELECT MD5_HEX(col) AS hex_hash, MD5_BINARY(col) AS binary_hash FROM table"
+        )
```
Comment (georgesittas) [inline tests/dialects/test_snowflake.py]: These two are redundant, let's remove them.

File: tests/fixtures/optimizer/annotate_functions.sql [hunk 7]
```diff
@@ -1935,6 +1935,46 @@ VARCHAR;
...
 VARCHAR;
 
+# dialect: snowflake
+MD5(tbl.str_col);
+VARCHAR;
+
+# dialect: snowflake
+MD5(NULL);
+VARCHAR;
+
```
Comment (georgesittas) [inline tests/fixtures/optimizer/annotate_functions.sql]: Btw, we don't have to add tests with `NULL` arguments in general. These are not really testing any special code paths.

**Review by georgesittas — APPROVED** [review]
Looks good, nice work. Just a [single ask](https://github.com/tobymao/sqlglot/pull/5883#discussion_r2350238580) from me before merging. The rest of the comments are just for cleaning up tests.
